provider/console/watcher.go

```diff
@@ -0,0 +1,126 @@
+package console
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/src-d/lookout"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-log.v1"
+)
+
+// Provider is the name
+const Provider = "console"
+
+// WatchOptions options to use in the Watcher constructor.
+type WatchOptions struct {
+	Reader io.Reader
+}
+
+// Watcher watches for new events in the console
+type Watcher struct {
+	o       *WatchOptions
+	scanner *bufio.Scanner
+}
+
+// NewWatcher returns a new console watcher
+func NewWatcher(o *WatchOptions) (*Watcher, error) {
+	return &Watcher{
+		o:       o,
+		scanner: bufio.NewScanner(o.Reader),
+	}, nil
+}
+
+// Watch reads from stdin and calls cb for each new event
+func (w *Watcher) Watch(ctx context.Context, cb lookout.EventHandler) error {
+	log.With(log.Fields{"provider": Provider}).Infof("Starting watcher")
+
+	lines := make(chan string, 1)
+	go func() {
+		for w.scanner.Scan() {
+			lines <- w.scanner.Text()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case line := <-lines:
+			if err := w.handleInput(cb, line); err != nil {
+				if lookout.NoErrStopWatcher.Is(err) {
+					return nil
+				}
+
+				return err
+			}
+		}
+	}
+}
+
+func (w *Watcher) handleInput(cb lookout.EventHandler, line string) error {
+	if line == "" {
+		return nil
+	}
+
+	var cmd, gitURL, fromBranch, fromHash, toBranch, toHash string
+	_, err := fmt.Sscanf(line,
+		"%s %s %s %s %s %s",
+		&cmd, &gitURL, &fromBranch, &fromHash, &toBranch, &toHash)
+
+	if err != nil {
+		log.Errorf(err, "could not process input line %q", line)
+		return nil
+	}
+
+	var event lookout.Event
+
+	commitRev := lookout.CommitRevision{
+		Base: lookout.ReferencePointer{
+			InternalRepositoryURL: gitURL,
+			ReferenceName:         plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", fromBranch)),
+			Hash:                  fromHash,
+		},
+		Head: lookout.ReferencePointer{
+			InternalRepositoryURL: gitURL,
+			ReferenceName:         plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", toBranch)),
+			Hash:                  toHash,
+		},
+	}
+
+	switch strings.ToLower(cmd) {
+	case "review":
+		event = &lookout.ReviewEvent{
+			Provider:    Provider,
+			InternalID:  "1",
```

Comment:
Any reason not to generate a random UUIDs here?

Reply:
If the point is to avoid repetition of the ID, a random ID may repeat itself, and to avoid it we would need to do something too complex for this provider (like keeping track of used IDs, even between restarts when we have a DB).

I'd argue that if repeating the ID is a problem at some point in the future, a fixed "1" will be detected quickly, and at that point we can figure something out.

Reply:
Random UUID has very low probability of being repeated and that is something (more or less) that is expected from other providers

Reply:
With the more flexible approach in #61 you can set "internal_id" manually for each event, so I think this is no longer an issue.